modules/HardwareDrivers/PowerMeters/CarloGavazzi_DCT1/meter/powermeterImpl.cpp

```diff
@@ -0,0 +1,192 @@
+        this->m_public_key_length_in_bits = 256;
+        break;
+    case SIGNATURE_384_BIT:
+        this->m_public_key_length_in_bits = 384;
+        break;
+    default:
+        throw std::runtime_error("no signature keys are configured");
+    }
+
+    this->m_public_key_hex = read_public_key(this->m_public_key_length_in_bits);
+
+    EVLOG_info << "Public key: " << this->m_public_key_hex.value() << "\n";
+}
+
+void powermeterImpl::ready() {
+    read_signature_config();
+
+    std::thread t([this] {
+        while (true) {
+            read_powermeter_values();
+            sleep(1);
+        }
+    });
+    t.detach();
+}
+
+types::powermeter::TransactionStartResponse
+powermeterImpl::handle_start_transaction(types::powermeter::TransactionReq& value) {
+    try {
+        // lock device mutex
+        std::scoped_lock lock(m_device_mutex);
+
+        this->m_start_signed_meter_value = read_signed_meter_values();
+
+        return {types::powermeter::TransactionRequestStatus::OK};
+    } catch (const std::exception& e) {
+        EVLOG_error << __PRETTY_FUNCTION__ << " Error: " << e.what() << std::endl;
+        return {types::powermeter::TransactionRequestStatus::UNEXPECTED_ERROR, {}, {}, "get_signed_meter_value_error"};
+    }
+}
+
+types::powermeter::TransactionStopResponse powermeterImpl::handle_stop_transaction(std::string& transaction_id) {
+    try {
+        // lock device mutex
+        std::scoped_lock lock(m_device_mutex);
+
+        auto signed_meter_value = read_signed_meter_values();
+
+        return {types::powermeter::TransactionRequestStatus::OK, this->m_start_signed_meter_value, signed_meter_value};
+    } catch (const std::exception& e) {
+        EVLOG_error << __PRETTY_FUNCTION__ << " Error: " << e.what() << std::endl;
+        return {types::powermeter::TransactionRequestStatus::UNEXPECTED_ERROR, {}, {}, "get_signed_meter_value_error"};
+    }
+}
+
+types::units_signed::SignedMeterValue powermeterImpl::read_signed_meter_values() {
+    int registers =
+        MODBUS_SIGNED_MAP_SIGNATURE_ADDRESS - MODBUS_SIGNED_MAP_ADDRESS + (this->m_public_key_length_in_bits >> 3);
+    transport::DataVector data = p_modbus_transport->fetch(MODBUS_SIGNED_MAP_ADDRESS, registers);
+    auto signed_meter_value = types::units_signed::SignedMeterValue{
+        /* signed_meter_data */ to_base64_string(data, 0, data.size()),
+        /* signing_method */
+        fmt::format("{0} bit ECDSA SHA {0}, using curve brainpoolP{0}r1", this->m_public_key_length_in_bits),
+        /* encoding_method - ?????? */ "",
+        /* public_key */ this->m_public_key_hex};
+    return signed_meter_value;
+}
+
+void powermeterImpl::read_powermeter_values() {
+    try {
+        // lock device mutex
+        std::scoped_lock lock(m_device_mutex);
+
+        transport::DataVector data = p_modbus_transport->fetch(MODBUS_REAL_TIME_VALUES_ADDRESS, 6);
+
+        float sign = 1.;
+        if (not config.low_side_sensing) {
+            sign = -1.;
```

Comment:
If it's explicitly `float`, might as well use the `f` suffix. (But the preprocessor will fix it anyway.)
Or make it int, it can be used in the multiplication below as well.
```suggestion
        float sign = 1.f;
        if (not config.low_side_sensing) {
            sign = -1.f;
```